### PR TITLE
Output global_attentions in Longformer models

### LINKED ISSUE #7514: [Longformer] Output both local attentions and global attentions when `output_attentions=True` -> Good Second Issue
# 🚀 Feature request

**Good Second Issue** - A more advanced issue for contributors who want to dive more into Longformer's attention mechanism.

Longformer currently only outputs global attentions, which is suboptimal because users might be interested in the local attentions as well. I propose to change the "output_attention" logic as follows in longformer:

`attentions` should correspond to the "local" attentions and then we'll add a new output type `global_attention` that contains the global_attentions. This is consistent with the naming of `attention_mask` and `global_attention_mask` IMO and the cleanest way to implement the feature.

Implementing this feature would mean to that Longformer will require its own `ModelOutput` class => 
`BaseModelOutput,` => `LongformerBaseModelOutput` or `BaseModelOutputWithGlobalAttention` (prefer the first name though)
`BaseModelOutputWithPooling,` => ...

Also some tests will have to be adapted.

This is a slightly more difficult issue, so I'm happy to help on it. One should understand the difference between local and global attention and how Longformer's attention is different to *e.g.* Bert's attention in general.

For more detail check out discussion here: https://github.com/huggingface/transformers/issues/5646

### LINKED ISSUE #8185: TensorFlow Longformer model as a saved model with attention outputs
The TensorFlow implementation of the Longformer model has an issue with using the `tf.saved_model.save` API alongside the `output_attentions=True` configuration attribute. 

The test is skipped currently due to this issue.

### PR DESCRIPTION
**Pull request by gui11aume**
# What does this PR do?

Fixes #7514

[From @patrickvonplaten]:

This PR introduces a new structure for the output attentions in Longformer. There are two types of attentions in Longformer: local attention outputs and global attention outputs. Previously, if global attention was used the `attentions` were set to the global attentions and the local attentions were discarded. This is suboptimal as one has no access to the local attentions in this case. The better design IMO is to have both `attentions` and `global_attentions` in Longformer (similar to `encoder_attentions`, `decoder_attentions` in Seq2Seq and `attentions`, `ngram_attentions` in ProphetNet).

Also, the PR switches from tuple indexing to using `ModelOutput` kwargs in the `test_attention_output` function which IMO we should bit by bit do for all tests from now on. In PT Longformer, the `is_global_attn_index` tensor is now only calculated once instead for each layer which slightly speeds up computation.

Awesome job @gui11aume! Especially for the docstring -> the description of `global_attentions` and `attentions` is impeccable.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dimiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), 
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests? 

## Who can review?

@patrickvonplaten

### CODE REVIEW

File: src/transformers/modeling_longformer.py [hunk 1]
```diff
@@ -63,6 +64,207 @@
...
 
+@dataclass
+class LongformerBaseModelOutput(ModelOutput):
+    """ 
+    Base class for Longformer's outputs, with potential hidden states, local and global attentions.
+
+    Args:
+        last_hidden_state (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_length, hidden_size)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+        hidden_states (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_hidden_states=True`` is passed or when ``config.output_hidden_states=True``):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py:75]:
```suggestion
        hidden_states (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when :obj:`output_hidden_states=True` is passed or when :obj:`config.output_hidden_states=True`):
```

File: src/transformers/modeling_longformer.py [hunk 2]
```diff
@@ -63,6 +64,207 @@
...
+            Tuple of :obj:`torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer)
+            of shape :obj:`(batch_size, sequence_length, hidden_size)`.
+
+            Hidden-states of the model at the output of each layer plus the initial embedding outputs.
+        attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py:80]:
```suggestion
        attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when :obj:`output_attentions=True` is passed or when :obj:`config.output_attentions=True`):
```

**Review by patrickvonplaten — COMMENTED** [review]
@gui11aume - ran the `make style` command so that the tests are all green. I will take a deeper look tomorrow :-) 
The approach looks exactly like I was planning on implementing it. I'll run the slow tests tomorrow and I might do some minor changes to your PR, but overall this looks great!

**Comment by patrickvonplaten** [comment]
We are actually having a longer internal discussion about the general handling of different attentions - this might still take a couple of days to be decided.

File: src/transformers/modeling_longformer.py [hunk 3]
```diff
@@ -96,7 +90,7 @@ class LongformerBaseModelOutput(ModelOutput):
             be accessed from 'global_attentions'.
         global_attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
             Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape
-            :obj:`(batch_size, num_heads, x, sequence_length)`, where `x` is the number of tokens with global
+            :obj:`(batch_size, num_heads, sequence_length, x)`, where `x` is the number of tokens with global
```
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py:93]: I changed the ordering here - I think it's better to be consistent with `attentions` in that first dim is the whole `seq_len` and the second dimension is `x` or `x + attn_window + 1`.

File: src/transformers/modeling_longformer.py [hunk 4]
```diff
@@ -440,13 +431,6 @@ def forward(
             +ve: global attention
 
         """
-        attention_mask = attention_mask.squeeze(dim=2).squeeze(dim=1)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_longformer.py:241]: This actually does not have to be re-computed for every layer, so I moved it to the `LongformerEncoder`. It could give a (very) small speed-up

File: src/transformers/modeling_longformer.py [hunk 5]
```diff
@@ -63,6 +58,194 @@
...
+            index is set to 0; the value should be accessed from the first `x` attention weights. If a token has global
+            attention, the attention weights to all other tokens in `attentions` is set to 0, the values should be
+            accessed from 'global_attentions'.
+        global_attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape :obj:`(batch_size, num_heads,
+            sequence_length, x)`, where `x` is the number of tokens with global attention mask.
+
+            Global attentions weights after the attention softmax, used to compute the weighted average in the
+            self-attention heads. Those are the attention weights from every token with global attention to every token
+            in the sequence.
```
Comment (LysandreJik) [inline src/transformers/modeling_longformer.py:142]: Really cool docstrings! You're putting all the variables in between ``, but this will only put them in *italics* in Sphinx, rather than the `monospace` I believe you're looking for? You can do so with double backticks or by preceding the variable between single backticks with :obj:

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -220,12 +220,13 @@ def test_attention_outputs(self):
...
             inputs_dict["output_attentions"] = True
             inputs_dict["output_hidden_states"] = False
+            config.return_dict = True
             model = model_class(config)
             model.to(torch_device)
             model.eval()
             with torch.no_grad():
                 outputs = model(**self._prepare_for_class(inputs_dict, model_class))
-            attentions = outputs[-1]
+            attentions = outputs.encoder_attentions if config.is_encoder_decoder else outputs.attentions
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:229]: Way better!

**Review by LysandreJik — APPROVED** [review]
Cool, wonderful that you added so many tests!

@patrickvonplaten, you say:

> if global attention was used the attentions were set to the global attentions and the local attentions were discarded

In this case, wouldn't the best model output be one where there are both `local_attentions`, `global_attentions` **as well as** `attentions` that are kept simply for backwards compatibility?

Other than that and the docstrings, LGTM!

File: src/transformers/modeling_longformer.py [hunk 7]
```diff
@@ -63,6 +58,194 @@
...
+        attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape :obj:`(batch_size, num_heads,
+            sequence_length, x + attention_window + 1)`, where `x` is the number of tokens with global attention mask.
+
+            Local attentions weights after the attention softmax, used to compute the weighted average in the
+            self-attention heads. Those are the attention weights from every token in the sequence to every token with
+            global attention (first `x` values) and to every token in the attention window (remaining `attention_window
+            + 1` values). Note that the first `x` values refer to tokens with fixed positions in the text, but the
+            remaining `attention_window + 1` values refer to tokens with relative positions: the attention weight of a
+            token to itself is located at index `x + `attention_window / 2` and the `attention_window / 2` preceding
```
Comment (sgugger) [inline src/transformers/modeling_longformer.py:130]:
```suggestion
            token to itself is located at index `x + attention_window / 2` and the `attention_window / 2` preceding
```
This is repeated in all other output docstrings.

File: src/transformers/modeling_tf_longformer.py [hunk 8]
```diff
@@ -53,6 +56,146 @@
...
+        attentions (:obj:`tuple(tf.Tensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
+            Tuple of :obj:`tf.Tensor` (one for each layer) of shape :obj:`(batch_size, num_heads, sequence_length, x +
+            attention_window + 1)`, where `x` is the number of tokens with global attention mask.
+
+            Local attentions weights after the attention softmax, used to compute the weighted average in the
+            self-attention heads. Those are the attention weights from every token in the sequence to every token with
+            global attention (first `x` values) and to every token in the attention window (remaining `attention_window
+            + 1` values). Note that the first `x` values refer to tokens with fixed positions in the text, but the
+            remaining `attention_window + 1` values refer to tokens with relative positions: the attention weight of a
+            token to itself is located at index `x + `attention_window / 2` and the `attention_window / 2` preceding
```
Comment (sgugger) [inline src/transformers/modeling_tf_longformer.py:81]:
```suggestion
            token to itself is located at index `x + attention_window / 2` and the `attention_window / 2` preceding
```
(and like before this is in all docstrings)

**Review by sgugger — APPROVED** [review]
This looks way cleaner, I like this a lot! There are small nits in the docstrings to address, otherwise it looks good to merge.

**Comment by patrickvonplaten** [comment]
> Cool, wonderful that you added so many tests!
> 
> @patrickvonplaten, you say:
> 
> > if global attention was used the attentions were set to the global attentions and the local attentions were discarded
> 
> In this case, wouldn't the best model output be one where there are both `local_attentions`, `global_attentions` **as well as** `attentions` that are kept simply for backwards compatibility?

The previous design led to errors as shown here: https://github.com/huggingface/transformers/issues/5646 -> so I think it's fine to break backwards compatibility here. `local_attentions` would arguably be a better name than `attentions`, but for consistency with other models and for a standard case where `global_attention_mask=None`, so that `local_attentions` == (all) `attentions`, I would prefer to keep the name `attentions` here. 

> 
> Other than that and the docstrings, LGTM!

Docstrings will be corrected!

**Comment by LysandreJik** [comment]
Works for me!

**Comment by patrickvonplaten** [comment]
@gui11aume great work again - understanding longformer's attention is not straightforward and your doc string was spot-on! Looking forward to your next contribution ;-) Hope you're fine with the small changes I made

**Comment by manishiitg** [comment]
@patrickvonplaten @lalitpagaria this merge PR breaks longformer training with gradient checkpointing True
please fix as i am unable to train models with latest models. 


error comes on this line 
https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py#L1072

model expect 2 to 6 positional arguments, but 7 where giving.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten @lalitpagaria this merge PR breaks longformer training with gradient checkpointing True
> please fix as i am unable to train models with latest models.
> 
> error comes on this line
> https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py#L1072
> 
> model expect 2 to 6 positional arguments, but 7 where giving.

Hey @manishiitg,

Thanks a lot for message! This did indeed break gradient checkpointing for longformer - sorry! This PR fixes it: https://github.com/huggingface/transformers/pull/8415
